### PR TITLE
Allow phase items not associated with meshes to be binned.

### LINKED ISSUE #14004: Can't upgrade to 0.14, impossible to render `AlphaMask3d` phase items without mesh
## Bevy version

`main`

## What you did

Repro branch : https://github.com/djeedai/bevy_hanabi/tree/bevy_main

Run:

```
cargo --offline r --example billboard --no-default-features --features="bevy/zstd bevy/ktx2 bevy/bevy_winit bevy/bevy_pbr bevy/png 3d "
```

The billboard particles are missing. GPU debugger shows there's no draw call.

Discussion on [Discord](https://discord.com/channels/691052431525675048/743663924229963868/1254458619185991700) and comments on related [migration guide bug](https://github.com/bevyengine/bevy-website/issues/1386).

Hanabi (`bevy_hanabi`) renders some particle effect through the `AlphaMask3d` pass via a custom render pipeline based on compute shaders and vertex/fragment shaders. There's no mesh involved (except from a single quad shared by all particles, which could even be removed). After migrating, I noted that `AlphaMask3d` uses `BinnedRenderPhase`, unlike all other phases Hanabi uses which go through `SortedRenderPhase`. The design of the 2 is quite different, and it seems `BinnedRenderPhase` is coupled with mesh asset rendering.

**This makes it impossible to ugprade Hanabi.**

## What went wrong

Render phase items `AlphaMask3d` are silently discarded, because (I think) `GetFullBatchData::get_binned_index()` returns `None`, since the phase item enqueued doesn't have a mesh.
- I've confirmed with a debugger the phase items get enqueued
- I've confirmed with a debugger that `gpu_processing::batch_and_prepare_binned_render_phase()` gets called (I _think_, because it's all generic code and I can't see the concrete types)

### PR DESCRIPTION
**Pull request by pcwalton**
As reported in #14004, many third-party plugins, such as Hanabi, enqueue entities that don't have meshes into render phases. However, the introduction of indirect mode added a dependency on mesh-specific data, breaking this workflow. This is because GPU preprocessing requires that the render phases manage indirect draw parameters, which don't apply to objects that aren't meshes. The existing code skips over binned entities that don't have indirect draw parameters, which causes the rendering to be skipped for such objects.

To support this workflow, this commit adds a new field, `non_mesh_items`, to `BinnedRenderPhase`.  This field contains a simple list of (bin key, entity) pairs. After drawing batchable and unbatchable objects, the non-mesh items are drawn one after another. Bevy itself doesn't enqueue any items into this list; it exists solely for the application and/or plugins to use.

Additionally, this commit switches the asset ID in the standard bin keys to be an untyped asset ID rather than that of a mesh. This allows more flexibility, allowing bins to be keyed off any type of asset.

This patch adds a new example, `custom_phase_item`, which simultaneously serves to demonstrate how to use this new feature and to act as a regression test so this doesn't break again.

Fixes #14004.

### Added

* `BinnedRenderPhase` now contains a `non_mesh_items` field for plugins to add custom items to.

### CODE REVIEW

File: assets/shaders/custom_phase_item.wgsl [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+struct Vertex {
+    // The world-space position of the vertex.
+    @location(0) position: vec3<f32>,
+    // The color of the vertex.
+    @location(1) color: vec3<f32>,
+};
+
+// Information passed from the vertex shader to the fragment shader. (The name
+// comes from OpenGL.)
+struct Varyings {
```
Comment (IceSentry) [inline assets/shaders/custom_phase_item.wgsl:16]: I think this should be either `FragmentInput` or `VertexOutput` for consistency with the other examples and bevy's internals.

**Review by IceSentry — APPROVED** [review]
I haven't done any testing yet, but the code and overall approach looks good. I'll do some testing tomorrow to make sure it didn't break anything unexpected.

File: examples/shader/custom_phase_item.rs [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        ibo.write_buffer(render_device, render_queue);
+
+        CustomPhaseItemBuffers {
+            vertices: vbo,
+            indices: ibo,
+        }
+    }
+}
+
+impl Vertex {
```
Comment (IceSentry) [inline examples/shader/custom_phase_item.rs:401]: This doesn't really matter so up to you if you want to do anything. I'm personally not a fan of having an impl block so far from the struct. It was a bit surprising to see it so far down.

Arguably, it doesn't matter for the example so it's fine to be down there and not pollute the start of the file which is why I don't mind if it stays there.

**Review by tychedelia — APPROVED** [review]
Looks great, ty! Just some nits on the example. Just one request: could you also try to add visibility logic with `check_visibility` for the non-mesh? It would be great to document this in the examples as it's another gotcha.

**Comment by superdump** [comment]
Before reviewing, the only thing that comes to mind is - can asset ids of different types collide when compared as untyped asset ids? I think they can. In that case, I think keying on both mesh assets and other assets could cause such items to be binned together when they shouldn’t be?

**Comment by pcwalton** [comment]
@superdump I modified the `PartialOrd` impl on `UntypedAssetId` to address this.

File: examples/shader/custom_phase_item.rs [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+            multisample: MultisampleState {
+                count: msaa.samples(),
+                mask: !0,
+                alpha_to_coverage_enabled: false,
+            },
+        }
+    }
+}
+
+impl<P> RenderCommand<P> for SetCustomPhaseItemPipeline
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply examples/shader/custom_phase_item.rs]: Yes, fixed.

**Comment by pcwalton** [comment]
Review comments addressed.

**Comment by djeedai** [comment]
I'll look later but the Changelog looks wrong, again `BinnedRenderPhase` didn't exist in 0.13 so we can't reference it. I think it would make more sense to collate with the release notes and Changelog of the original feature adding bin/sort phases.

**Comment by pcwalton** [comment]
The problem is there's no way in the pull request message to apply a patch to an existing changelog entry. I could delete the changelog entry if you want.

**Comment by tychedelia** [comment]
Tbh, I think having the incremental change log is helpful for developers who build of main. I've identified and upgrading things myself this way in between release cycles. Perhaps we just need to disambiguate change fields better.

File: crates/bevy_core_pipeline/src/core_3d/mod.rs [hunk 4]
```diff
@@ -223,7 +222,7 @@ pub struct Opaque3dBinKey {
     pub draw_function: DrawFunctionId,
 
     /// The mesh.
```
Comment (djeedai) [inline crates/bevy_core_pipeline/src/core_3d/mod.rs:224]:
```suggestion
    /// The asset to batch the phase items by.
```

**Review by djeedai — APPROVED** [review]

**Pull request merged**